sql/catalyst/src/main/scala/org/apache/spark/sql/types/DecimalType.scala

```diff
@@ -76,23 +77,29 @@ case class DecimalType(precision: Int, scale: Int) extends FractionalType {
    * Returns whether this DecimalType is wider than `other`. If yes, it means `other`
    * can be casted into `this` safely without losing any precision or range.
    */
-  private[sql] def isWiderThan(other: DataType): Boolean = other match {
+  private[sql] def isWiderThan(other: DataType): Boolean = isWiderThanInternal(other)
+
+  @tailrec
+  private def isWiderThanInternal(other: DataType): Boolean = other match {
     case dt: DecimalType =>
       (precision - scale) >= (dt.precision - dt.scale) && scale >= dt.scale
     case dt: IntegralType =>
-      isWiderThan(DecimalType.forType(dt))
+      isWiderThanInternal(DecimalType.forType(dt))
     case _ => false
   }
 
   /**
    * Returns whether this DecimalType is tighter than `other`. If yes, it means `this`
    * can be casted into `other` safely without losing any precision or range.
    */
-  private[sql] def isTighterThan(other: DataType): Boolean = other match {
+  private[sql] def isTighterThan(other: DataType): Boolean = isTighterThanInternal(other)
```

Comment:
Otherwise:

```
[error] /.../spark/sql/catalyst/src/main/scala/org/apache/spark/sql/types/DecimalType.scala:94:20: could not optimize @tailrec annotated method isTighterThan: it is neither private nor final so can be overridden
[error]   private[sql] def isTighterThan(other: DataType): Boolean = other match {
[error]                    ^
```